File: parsers/typescript/test/parse.test.ts
import { parseSponsorTextVariables } from "../lib/parse"

describe("parseSponsorTextVariables", () => {
    test("parses variables with suffix", () => {
        // GIVEN
        const text = "$vars:discord=https://discord.gg/vx4AKRfj;provider=bf2hub.com$"

        // WHEN
        const variables = parseSponsorTextVariables(text)

        expect(variables).toEqual({
            "discord": "https://discord.gg/vx4AKRfj",
            "provider": "bf2hub.com",
        })
    })

    test("parses variables without suffix", () => {
        // GIVEN
        const text = "$vars:discord=https://discord.gg/vx4AKRfj;provider=bf2hub.com"

        // WHEN
        const variables = parseSponsorTextVariables(text)

        expect(variables).toEqual({
            "discord": "https://discord.gg/vx4AKRfj",
            "provider": "bf2hub.com",
        })
    })

    test("skips non-variables prefix", () => {
        // GIVEN
        const text = "Join our event this Sunday! $vars:discord=https://discord.gg/vx4AKRfj;provider=bf2hub.com$"

        // WHEN
        const variables = parseSponsorTextVariables(text)

        expect(variables).toEqual({
            "discord": "https://discord.gg/vx4AKRfj",
            "provider": "bf2hub.com",
        })
    })

    test("stops before non-variables suffix", () => {
        // GIVEN
        const text = "$vars:discord=https://discord.gg/vx4AKRfj;provider=bf2hub.com$ Apply to become an admin today!"

        // WHEN
        const variables = parseSponsorTextVariables(text)

        expect(variables).toEqual({
            "discord": "https://discord.gg/vx4AKRfj",
            "provider": "bf2hub.com",
        })
    })

    test("reads escaped syntax characters", () => {
        // GIVEN
        const text = "$vars:\\$trange\\=key=https://example.com?query\\=\\$start\\;end$"

        // WHEN
        const variables = parseSponsorTextVariables(text)

        expect(variables).toEqual({
            "$trange=key": "https://example.com?query=$start;end",
        })
    })

    test("reads escaped whitespaces", () => {
        // GIVEN
        const text = "$vars:four-spaces=\\ \\ \\ \\ $"

        // WHEN
        const variables = parseSponsorTextVariables(text)

        expect(variables).toEqual({
            "four-spaces": "    ",
        })
    })

    test("omits irrelevant whitespaces", () => {
        // GIVEN
        const text = "$vars: trimmed = one  two  three $"

        // WHEN
        const variables = parseSponsorTextVariables(text)

        expect(variables).toEqual({
            "trimmed": "one two three",
        })
    })

    test("ignores incomplete key", () => {
        // GIVEN
        const text = "$vars:discord=https://discord.gg/vx4AKRfj;website$"

        // WHEN
        const variables = parseSponsorTextVariables(text)

        expect(variables).toEqual({
            "discord": "https://discord.gg/vx4AKRfj",
        })
    })

    test("includes key with no value", () => {
        // GIVEN
        const text = "$vars:discord=https://discord.gg/vx4AKRfj;teamspeak=$"

        // WHEN
        const variables = parseSponsorTextVariables(text)

        expect(variables).toEqual({
            "discord": "https://discord.gg/vx4AKRfj",
            "teamspeak": "",
        })
    })

    test("returns empty object for empty variables section", () => {
        // GIVEN
        const text = "$vars:$"

        // WHEN
        const variables = parseSponsorTextVariables(text)

        expect(variables).toEqual({})
    })

    test("returns empty object for sponsor text without variables", () => {
        // GIVEN
        const text = "Our server is the best!"

        // WHEN
        const variables = parseSponsorTextVariables(text)

        expect(variables).toEqual({})
    })
})

File: parsers/go/internal/parse_test.go
package internal_test

import (
	"testing"

	"github.com/stretchr/testify/assert"

	"sponsortext/internal"
)

func TestParseSponsorTextVariables(t *testing.T) {
	tests := []struct {
		name  string
		given string
		want  map[string]string
	}{
		{
			name:  "parses variables with suffix",
			given: "$vars:discord=https://discord.gg/vx4AKRfj;provider=bf2hub.com$",
			want: map[string]string{
				"discord":  "https://discord.gg/vx4AKRfj",
				"provider": "bf2hub.com",
			},
		},
		{
			name:  "parses variables without suffix",
			given: "$vars:discord=https://discord.gg/vx4AKRfj;provider=bf2hub.com",
			want: map[string]string{
				"discord":  "https://discord.gg/vx4AKRfj",
				"provider": "bf2hub.com",
			},
		},
		{
			name:  "skips non-variables prefix",
			given: "Join our event this Sunday! $vars:discord=https://discord.gg/vx4AKRfj;provider=bf2hub.com$",
			want: map[string]string{
				"discord":  "https://discord.gg/vx4AKRfj",
				"provider": "bf2hub.com",
			},
		},
		{
			name:  "stops before non-variables suffix",
			given: "$vars:discord=https://discord.gg/vx4AKRfj;provider=bf2hub.com$ Apply to become an admin today!",
			want: map[string]string{
				"discord":  "https://discord.gg/vx4AKRfj",
				"provider": "bf2hub.com",
			},
		},
		{
			name:  "reads escaped syntax characters",
			given: "$vars:\\$trange\\=key=https://example.com?query\\=\\$start\\;end$",
			want: map[string]string{
				"$trange=key": "https://example.com?query=$start;end",
			},
		},
		{
			name:  "reads escaped whitespaces",
			given: "$vars:four-spaces=\\ \\ \\ \\ $",
			want: map[string]string{
				"four-spaces": "    ",
			},
		},
		{
			name:  "omits irrelevant whitespaces",
			given: "$vars: trimmed = one  two  three ; stri pped=four $",
			want: map[string]string{
				"trimmed":  "one two three",
				"stripped": "four",
			},
		},
		{
			name:  "ignores incomplete key",
			given: "$vars:discord=https://discord.gg/vx4AKRfj;website$",
			want: map[string]string{
				"discord": "https://discord.gg/vx4AKRfj",
			},
		},
		{
			name:  "includes key with no value",
			given: "$vars:discord=https://discord.gg/vx4AKRfj;teamspeak=$",
			want: map[string]string{
				"discord":   "https://discord.gg/vx4AKRfj",
				"teamspeak": "",
			},
		},
		{
			name:  "returns empty map for empty variables section",
			given: "$vars:$",
			want:  map[string]string{},
		},
		{
			name:  "returns nil for sponsor text without variables",
			given: "Our server is the best!",
			want:  nil,
		},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			// WHEN
			variables := internal.ParseSponsorTextVariables(tt.given)

			// THEN
			assert.Equal(t, tt.want, variables)
		})
	}
}
